src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -0,0 +1,145 @@
+
+    private static final Gson GSON = new Gson();
+    private final InternalConnection connection;
+    private final Authorizer authorizer;
+
+    private class PrivateEncryptedChannelData {
+        final String auth;
+        final String sharedSecret;
+        final String channelData;
+
+        protected PrivateEncryptedChannelData(String auth, String sharedSecret, String channelData) {
+            this.auth = auth;
+            this.sharedSecret = sharedSecret;
+            this.channelData = channelData;
+        }
+
+        public String getAuth() {
+            return auth;
+        }
+
+        public String getSharedSecret() {
+            return sharedSecret;
+        }
+
+        public String getChannelData() {
+            return channelData;
+        }
+    }
+
+    private PrivateEncryptedChannelData authorizerData;
+
+    public PrivateEncryptedChannelImpl(final InternalConnection connection,
+                                       final String channelName,
+                                       final Authorizer authorizer,
+                                       final Factory factory) {
+        super(channelName, factory);
+        this.connection = connection;
+        this.authorizer = authorizer;
+    }
+
+    /* PrivateChannel implementation */
+
+
+    /* Base class overrides */
+
+    @Override
+    public void bind(final String eventName, final SubscriptionEventListener listener) {
+
+        if (!(listener instanceof PrivateEncryptedChannelEventListener)) {
+            throw new IllegalArgumentException(
+                    "Only instances of PrivateEncryptedChannelEventListener can be bound " +
+                            "to a private encrypted channel");
+        }
+
+        super.bind(eventName, listener);
+    }
+
+    private void saveSharedSecret() {
+        final String authResponse = getAuthResponse();
+        final Map authResponseMap = GSON.fromJson(authResponse, Map.class);
+        final String sharedSecret = (String)authResponseMap.get("shared_secret");
+        final byte[] sharedSecretBase64 = Base64.getDecoder().decode(sharedSecret);
+
+        // todo set up secret box opener -> pass the key
+        // todo make sure when unsubscribe we clear the text
+        // todo can we clear everything when the user disconnects totally
+    }
+
+    /**
+     * ensure we've got all the bits and pieces we need to continue
+     */
+    protected void prepareChannel() {
+        final String authResponse = getAuthResponse();
+
+        try {
+            final Map authResponseMap = GSON.fromJson(authResponse, Map.class);
+            final String authKey = (String) authResponseMap.get("auth");
+            final String sharedSecret = (String) authResponseMap.get("shared_secret");
+            final String channelData = (String)authResponseMap.get("channel_data");
+
+            if (authKey == null || sharedSecret == null) {
```

Comment:
Marek and I had a discussion about removing the check on `if (authKey == null || sharedSecret == null) {` - it'll throw an error if we try to save a null key to the secret box and get caught in the other catch.

I think for now, i'd like to keep this check here because it's an explicit requirement to pass the next piece and will help inform developers exactly what's missing.

Reply:
Coming back to this, i think we can make this more descriptive e.g. if we're missing the auth token, or the shared secret we can tell people this in plain english.

Reply:
Have updated it to read `Didn't receive all the fields we expected  "from the Authorizer, expected an auth token and shared_secret but got: " + authResponse`.